frontend/catalyst/ag_primitives.py

```diff
@@ -349,28 +391,34 @@ def _call_python_while(loop_test, loop_body, get_state, _set_state, _nonlocals,
     return get_state()
 
 
-def while_stmt(loop_test, loop_body, get_state, set_state, nonlocals, symbol_names):
+def while_stmt(loop_test, loop_body, get_state, set_state, symbol_names, _opts):
     """An implementation of the AutoGraph 'while ..' statement. The interface is defined by
     AutoGraph, here we merely provide an implementation of it in terms of Catalyst primitives."""
 
     fallback = False
     init_state = get_state()
 
     try:
-        results = _call_catalyst_while(
-            loop_test, loop_body, get_state, set_state, nonlocals, symbol_names
-        )
+        results = _call_catalyst_while(loop_test, loop_body, get_state, set_state, symbol_names)
 
     except Exception as e:  # pylint: disable=broad-exception-caught
         if catalyst.autograph_strict_conversion:
             raise e
+
+        # On fallback to Python we need to remove the jax primitive for this loop, if it was
+        # queued. The reason is that it may be malformed which can cause issues during lowering.
+        if EvaluationContext.is_tracing():  # pragma: no branch
```

Comment:
@dime10 what are you planning to do with the quantum tape? If quantum tracing is enabled we are not going to have `qwhile` primitive in eqns (see my other comment with a test example). Instead there will be `ForLoop` on the quantum tape. Did you expect to handle these cases as well? 
Have you considered saving the contents of the `jax_frame.eqns` before going into loop and restoring it in case of exception? I think it would be more reliable and probably we could do the same for quantum tapes.

Reply:
Great comment, I forgot about the quantum tape! I like your idea, I wonder though whether copying the entire tape & jaxpr before each control flow primitive might not slow down the tracing ?

Reply:
I reimplemented the program reset to an earlier state without IDs. It should be much more robust now (although maybe not as robust as caching the full state every time), while still being just as efficient. It's based on the number of instructions at a particular point as a proxy for program state.

I also extended it to work in a quantum tracing context. 

https://github.com/PennyLaneAI/catalyst/pull/352/commits/c31d6afd5e8e6cc5b0bcfdced86ba773239aa2d9